pkg/kubevirt/data_volume.go

```diff
@@ -0,0 +1,118 @@
+package kubevirt
+
+import (
+	"context"
+	"fmt"
+
+	clouderrors "github.com/gardener/gardener-extension-provider-kubevirt/pkg/errors"
+
+	v1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog"
+	cdi "kubevirt.io/containerized-data-importer/pkg/apis/core/v1alpha1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// DataVolumeManager manages Kubevirt DataVolume operations.
+type DataVolumeManager interface {
+	// CreateVolume creates a new kubevirt Data Volume from the data volume specs and in the passed namespace.
+	CreateVolume(ctx context.Context, secret *v1.Secret, name, namespace string, dvSpec cdi.DataVolumeSpec) error
+	// GetVolume fetches the specified volume by the passed name and namespace and return DataVolumeNotFoundError error in case of
+	// not found object error.
+	GetVolume(ctx context.Context, secret *v1.Secret, name, namespace string) (*cdi.DataVolume, error)
+	// ListDataVolumes lists all the Data Volumes which exists in the passed namespace.
+	ListDataVolumes(ctx context.Context, secret *v1.Secret, namespace string) (*cdi.DataVolumeList, error)
+	// DeleteVolume delete the DataVolume based on the passed name and namespace.
+	DeleteVolume(ctx context.Context, secret *v1.Secret, name, namespace string) error
+}
+
+type defaultDataVolumeManager struct {
+	client ClientFunc
+}
+
+// NewDefaultDataVolumeManager creates a new default manager with a k8s controller-runtime based on the sent kubecomfig.
+func NewDefaultDataVolumeManager(client ClientFunc) (DataVolumeManager, error) {
+	return &defaultDataVolumeManager{
+		client: client,
+	}, nil
+}
+
+// CreateVolume creates a new kubevirt Data Volume from the data volume specs and in the passed namespace.
+func (d *defaultDataVolumeManager) CreateVolume(ctx context.Context, secret *v1.Secret, name, namespace string, dvSpec cdi.DataVolumeSpec) error {
+	dataVolume := &cdi.DataVolume{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+		},
+		Spec: dvSpec,
+	}
+
+	c, err := d.client(secret)
+	if err != nil {
+		return fmt.Errorf("failed to create kubevirt client: %v", err)
+	}
+	return c.Create(ctx, dataVolume)
```

Comment:
What will happen if the data volume has been already created? From my perspective, the call could be e.g. retried and should handle the case where the desired data volume already exists gracefully, without failing with an error.